node/src/components/block_validator.rs

```diff
@@ -524,7 +524,7 @@ impl BlockValidator {
     fn handle_transaction_fetched<REv>(
         &mut self,
         effect_builder: EffectBuilder<REv>,
-        dt_hash: DeployOrTransferHash,
+        transaction_hash: TransactionHash,
```

Comment:
Originally, we used `DeployOrTransferHash` to call a block `Invalid` when there was a stray deploy in the transfers set or vice versa.

Consider the following test, which fails:
```rust
/// Verifies detection of a transfer in transactions set.
#[tokio::test]
async fn stray_transfer_in_transaction_set() {
    let mut rng = TestRng::new();
    let ttl = TimeDiff::from_millis(200);
    let timestamp = Timestamp::from(1000);
    let txn_legacy = Transaction::from(new_legacy_deploy(&mut rng, timestamp, ttl));
    let txn_v1 = Transaction::from(new_v1_standard(&mut rng, timestamp, ttl));
    let transfer_legacy = Transaction::from(new_legacy_transfer(&mut rng, timestamp, ttl));

    let transactions = vec![
        transfer_legacy.clone(), // Doesn't belong here.
        txn_v1.clone(),
        txn_legacy.clone(),
    ];
    let transfers = vec![];
    assert!(!validate_block(&mut rng, timestamp, transactions, transfers, vec![], vec![]).await);
}
```

This issue should originally be detected by `transfer_deploy_mixup_and_replay()`test, however, there is a flaw in this test - it reuses the transactions/transfers, hence the validation fails at "[duplicated transaction in proposed block](https://github.com/casper-network/casper-node/blob/dev/node/src/components/block_validator/state.rs#L157)".

I [put up a PR](https://github.com/EdHastingsCasperAssociation/casper-node/pull/13) with updated test targeting your branch.

This is why I kept the `DeployOrTransferHash` in the original implementation and [was considering](https://github.com/casper-network/casper-node/pull/4513/files#diff-7b869e572fddfa9a5c44e9a48aff92884ac3e58cc8fbbbf0de7c8aa91b4e8ac2R541-R542) using [TypedTransactionHash](https://github.com/casper-network/casper-node/blob/feat-2.0/node/src/types/transaction/typed_transaction_hash.rs#L9) to allow similar kind of validation for the new transaction type.

This should not be a problem as long as malicious node is not considered, since we can theoretically trust that proposed block has the transactions bucketed correctly.

Reply:
@rafal-ch I think the test here is wrong, looking at the code, the block validator when it receives a proposed block, it will create an appendable block and check if the limits for any of the specific categories are hit, if any of the transaction categories break this limit, that is when we deem the block invalid. However, those limits are defined as chainspec values, and using the local chainspec values none of those limits are hit, and none of the production code apart from mentioned validation logic actually cares about whether its a mint, auction, standard or the install/upgrade. 

Additionally, I think the `transfer_deploy_mixup_and_replay()` needs to be corrected to only check for replays, because at the end of the day they are all transactions, and the notion of a `transfer` being mixed in with other `deploys` or `transactions` is incorrect

Reply:
That test does not exist on branch. The adding of it reveals a flaw in the mocked test logic, which is modeling something other than what the production code concerns itself with and then amusing itself by running tests against that mock. The actual production logic of the buffer, appendable block, and consensus payload manage all of this at a more abstract level...correctly so. It would be better to replace all of these baroque fixture based tests with more a more direct treatment that tests the actual production implementation. 

A few points: there was some kind of fundamental confusion in the original block validator logic which seemed to not understand that transfers _are_ deploys. The only time and place a distinction between them is relevant is when checking the count per category limit when putting together a proposable block and when validating a proposable block. The determination of which category a given deploy belonged in was and is a fact expressed on the deploy itself. 

This fundamental misunderstanding seems to have then multiplied further with the introduction of transaction v1, to the extent that some of this updated code was trying to track 6 vecs to distinguish between the various combinations of legacy deploy / new transaction kind. This is entirely unnecessary and belabors the logic with a lot of type munge. Given a transaction, regardless of whether it is a legacy deploy, a new transaction, or a future V2 V3 V9999 transaction, what category it is counted against is an explicit fact on the transaction itself. 

Where it matters is in the appendable block logic where it determines what category a transaction currently goes into, determines what the configured limit for that category is, and determines if adding one more instance of that kind would exceed that limit. This is all expressed in the one place in the add_transaction method of appendable block, used by the buffer on the sending side, and by the validation on the receiving side. This is the only place this determination needs to be made, and it is made on the fly against the summary of relevant information about the transaction being provided to that logic (aka the transaction footprint instance).

All the proposed block validation logic needs to prove is that if a proposed block we objectively know to be valid per the appendable block logic is received the validation logic should agree that it is valid, and if a proposed block we objectively know to be invalid per the appendable logic is received the validation logic should agree that it is invalid in precisely the way we know it to be. The rest is just wiring up explicit permutations of various valid and invalid block considerations. 

The complexity of different versions of transactions, different arrangements of block payloads, different configured block limits should not be leaked into all adjacent logic in a combinatorial explosion of types and juggling of corresponding collections of said types. Rather, the abstraction of Transaction should not leak allowing the vast majority of the logic to only care about Transaction and TransactionHash. The place where a block payload is produced and is verified per the current versions and limits in force at a particular point in time is singular and is the only place where the complexity is relevant and even that is easily abstracted by a taxonomic function of identify the kind...which is being handled now in the Transaction::footprint() smart constructor and the unified TransactionCategory. Also, given the recent tendency for people to extract types intended for internal implementation details out and making them public types, TransactionFootprint and TransactionCategory is for internal use as a convenience, should not be made public, and can be updated in future releases if we add or remove block categories or block limits requiring different distilled data points to make determinations about.